### PR TITLE
 Improvements for Prosody

### PR DESCRIPTION
**Pull request by florianjacob**
###### Motivation for this change
Rebased and conflict-resolved version of https://github.com/NixOS/nixpkgs/pull/27716 by @sshisk.

Could remove some commits due to package updates that already landed in master.
Additionally updated prosody community modules to current HEAD.

###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `build-use-sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [X] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [X] Tested execution of all binary files (usually in `./result/bin/`)
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

### CODE REVIEW

File: pkgs/servers/xmpp/prosody/default.nix
Comment (orivej) [inline pkgs/servers/xmpp/prosody/default.nix]: You should add assertions for the new optional features. `withExtraLibs` should probably be removed.

Comment (florianjacob) [reply pkgs/servers/xmpp/prosody/default.nix]: Good catch, thank you. I added an assert for `withDBI` now, which is the only new feature which needs one, I think.

Regarding `withExtraLibs`, I now reconstructed its intention: Some community modules have additional dependencies not stated in the package, like luapam or lualdap, which can be injected via `withExtraLibs = [ luapam ]` when `withCommunityModules = [ auth_pam ]` was added. I added a comment to withExtraLibs now.

@orivej Or is there another way to add those extra dependencies when configuring the package, can you do the same with overrideAttrs or something (I don't know, never needed that so far)? Then I'm quite sure it can be removed.

**Comment by florianjacob** [comment]
cc @flosse, didn't see the `pkgs/server/xmpp/prosody` part has a maintainer.

**Comment by flosse** [comment]
@florianjacob I'm sorry to discontinue as a maintainer.
I removed `flosse` from the lists: https://github.com/flosse/nixpkgs/tree/remove-flosse-as-maintainer

**Comment by florianjacob** [comment]
@flosse thank you for your time as maintainer, still. 👍 Should I pull your commits here in this branch, or will you pose a separate PR?

File: pkgs/servers/xmpp/prosody/default.nix
Comment (florianjacob) [reply pkgs/servers/xmpp/prosody/default.nix]: Tried to do some research on the topic. With `withExtraLibs`, you only need this to add community modules:
```
nixpkgs.config.packageOverrides = pkgs: {
  prosody = pkgs.prosody.override {
    withCommunityModules = [ "auth_pam" ];
    withExtraLibs = [ lua51Packages.luaposix ];
  };
};
```

Without it, you need to use `overrideAttrs` to create a custom package to add the new libs on top of the default ones,
as `libs` is no package option (if I understood it right). And then assign that custom package to the prosody service.
```
environment.systemPackages = [
  (pkgs.emacs.overrideAttrs (oldAttrs: {
    name = "prosody-with-mods";
    withCommunityModules = [ "auth_pam" ];
    libs = oldAttrs.libs ++ lua51Packages.luapam;
  }))
];
# hypothetical module option which does not exist yet
services.prosody.pkg = pkgs.prosody-with-mods;
```

So it could be made to work without that extra option, I guess.
But from a user's perspective, the first one looks much easier to me. But again, I have no experience with customizing packages.

**Comment by florianjacob** [comment]
@flosse merged your commits into this branch now. Thanks!

**Review by orivej — APPROVED** [review]

**Comment by florianjacob** [comment]
@fpletz now after @orivej has approved this, is there anything still missing from my side? 😃

**Comment by orivej** [comment]
Sorry for leaving this PR hanging, and thank you!

**Comment by florianjacob** [comment]
@orivej no worries! 🍻

**Pull request merged**
